[Security Solution] add prevalence expanded section to expandable flyout

PhilippeOberti opened the pull request:
## Summary

This PR adds the Prevalence details section under the Insights section of the left expanded panel of the expandable flyout.

The PR introduces the following main changes:
- the `PrevalenceDetails` placeholder component now retrieves the highlighted fields (using the same hook the `PrevalenceOverview` component was using, then displays a table with the calculated values
- 2 new `PrevalenceDetailsCountCell` and `PrevalenceDetailsPrevalenceCell` components, that are using within the `EuiBasicTable` cell `render` methods
- the above 2 cell components leverage a couple of hooks (`useFetchUniqueByField`, `useFetchFieldValuePairWithAggregation` and `useFetchFieldValuePairByEventType`) to fetch the data and do the necessary calculations. The hooks were very similar to the work done in the previous [prevalence overview PR](https://github.com/elastic/kibana/pull/158038) and were therefore made more generic and extracted to the `shared` folder at the root of the `flyout` folder. The hooks used within the `right` folder were replaced by these new generic ones

Some smaller changes were also made:
- `LeftFlyoutContext` was renamed to `LeftPanelContext` to be consistent with the `RightPanelContext` used in the other folder

Data is being fetched using ReactQuery.

### How to test

- add `xpack.securitySolution.enableExperimental: ['securityFlyoutEnabled']` to the `kibana.json` file
- run `yarn es snapshot --license trial`,  `yarn test:generate` and `yarn start --no-base-path`
- go to the Alerts page, and click on the expand detail button on any row of the table
- navigate to the Overview tab

### How to generate correlations data

See the description of the [previous PR](https://github.com/elastic/kibana/pull/158038) and [this comment](https://github.com/elastic/kibana/pull/158038#issuecomment-1553625788) to learn how to generate data.

![Screenshot 2023-05-26 at 4 45 16 PM](https://github.com/elastic/kibana/assets/17276605/26e89c08-164f-41d8-8a8c-73df785e287d)

https://github.com/elastic/security-team/issues/6423

## TODO

- [x] implement e2e tests
- [x] verify interfaces and functions documentations

PhilippeOberti committed:
[Security Solution] add prevalence expanded section to expandable flyout

christineweng commented on x-pack/plugins/security_solution/public/flyout/left/context.tsx:
```diff
@@ -29,16 +29,24 @@ export interface LeftPanelContext {
+  /**
+   * Retrieves searchHit values for the provided field
+   */
+  getFieldsData: (field: string) => unknown | unknown[];
```
instead of casting to unknown, consider reuse `GetFieldsData` type in `x-pack/plugins/security_solution/public/common/hooks/use_get_fields_data.ts`

PhilippeOberti replied:
yes thank you, not sure why I didn't use it. I remember you commenting on this on a previous PR also :D

christineweng commented on x-pack/plugins/security_solution/public/flyout/shared/utils/fetch_data.ts:
this is awesome ❤️

christineweng commented on x-pack/plugins/security_solution/public/flyout/shared/hooks/use_fetch_field_value_pair_by_event_type.ts:
```diff
@@ -0,0 +1,180 @@
+   */
+  values: string[];
+  /**
+   *
```
should this line have some text?

PhilippeOberti replied:
yes definitely, thanks!

christineweng commented on x-pack/plugins/security_solution/public/flyout/shared/hooks/use_fetch_field_value_pair_with_aggregation.test.ts:
```diff
@@ -0,0 +1,90 @@
+    },
+  });
+  (useDeepEqualSelector as jest.Mock).mockReturnValue({ to: '', from: '' });
+  (useGlobalTime as jest.Mock).mockReturnValue({ to: '', from: '' });
```
consider adding a test to check it is pulling the correct to and from based on isActiveTimelines

christineweng replied:
yeah that's overkill for sure. My comment was about both of these 2 sets of `from` and `to` are empty strings, it would appears that whether `isTimelines` is true of not would yield the same results. Maybe change the return value to `timelineTo` or something, and check that when `isTimelines` is true, the correct params are being used, does that make sense?

PhilippeOberti replied:
ok I finally got some time to look at this, and it's not straightforward. The `from` and `to` values are passed to the `buildSearchRequest` which isn't exported. Then the result of this is passed to `useQuery`.

Best way would be to export the line into its own hook but that definitely seems overkill to me. Other solution would be to export the `buildSearchRequest` method to be able to use it in the tests, but I'm not a fan of exporting things like that just for the sake of unit tests.

I'm tempted to not test this line, wdyt?

christineweng commented on x-pack/plugins/security_solution/public/flyout/left/components/prevalence_details.tsx:
```diff
@@ -5,17 +5,169 @@
+const columns: Array<EuiBasicTableColumn<unknown>> = [
+  {
+    field: 'type',
+    name: 'Type',
```
column titles need translation

christineweng reviewed (commented):
Desk tested and it looks great! 

I noticed the `(%)` in `Host Prevalence (%)` is cut off, unless i have kibana full screen AND on a ultra-wide monitor. In that case showing just numbers look a little off. I wonder if adding % to the numbers makes more sense to customers, wdyt?

![image](https://github.com/elastic/kibana/assets/18648970/97d2ea9c-6eeb-43bd-8c2b-27270af44e30)

PhilippeOberti committed:
PR comments:
- add doc
- lefft and right context more consistent
- add table column title translations
- more % to cells instead of column title

lgestc commented on x-pack/plugins/security_solution/cypress/e2e/investigations/alerts/expandable_flyout/alert_details_left_panel_prevalence_tab.cy.ts:
```diff
@@ -0,0 +1,74 @@
+  { env: { ftrConfig: { enableExperimental: ['securityFlyoutEnabled'] } } },
+  () => {
+    before(() => {
+      cleanKibana();
```
is cleaning kibana still required?

lgestc commented on x-pack/plugins/security_solution/public/flyout/left/components/analyze_graph.stories.tsx:
```diff
@@ -32,9 +31,9 @@ export default {
 //
 //   return (
-//     <LeftFlyoutContext.Provider value={contextValue}>
+//     <LeftPanelContext.Provider value={contextValue}>
 //       <AnalyzeGraph />
-//     </LeftFlyoutContext.Provider>
+//     </LeftPanelContext.Provider>
 //   );
```
why keep these commented?:)

PhilippeOberti replied:
good point, I've been thinking about deleting this file and the session view one for a while, so now is as good of a time as ever :)

lgestc commented on x-pack/plugins/security_solution/public/flyout/left/components/prevalence_details.test.tsx:
```diff
@@ -0,0 +1,82 @@
+} as unknown as LeftPanelContext;
+
+describe('PrevalenceDetails', () => {
+  (useFetchFieldValuePairByEventType as jest.Mock).mockReturnValue({
```
jest.mocked(useFetchFieldValue) would work better nowadays

PhilippeOberti replied:
I've converted most of them to `jest.mocked` but not all of them as we loose the fact that `as jest.Mock` ignores type check. In most scenarios it's good to have type check, but when we care only about a couple of properties on a giant object, using `as jest.Mock` is very convenient :)

lgestc commented on x-pack/plugins/security_solution/public/flyout/left/components/prevalence_details.test.tsx:
```diff
@@ -0,0 +1,82 @@
+    error: false,
+    count: 1,
+  });
+  (useFetchFieldValuePairWithAggregation as jest.Mock).mockReturnValue({
```
see above

lgestc commented on x-pack/plugins/security_solution/public/flyout/left/components/prevalence_details.tsx:
```diff
@@ -5,17 +5,177 @@
+    scopeId,
+  });
+
+  const data = (summaryRows || []).map((summaryRow) => {
+    const fields = getCellRenderFields(summaryRow);
+    return {
+      type: summaryRow.description.data.field,
+      name: summaryRow.description.values,
+      alertCount: fields,
+      docCount: fields,
+      hostPrevalence: fields,
+      userPrevalence: fields,
+    };
+  });
```
put this inside useMemo as well?:)

PhilippeOberti replied:
goot point, thanks for catching that!

lgestc commented on x-pack/plugins/security_solution/public/flyout/left/components/prevalence_details_count_cell.tsx:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  if (error) {
+    return null;
```
actually maybe we should return something here?:)

PhilippeOberti replied:
I originally thought that having an empty cell would be ok, but I agree that showing some sort of error is better. I decided to return an error icon, let me know what you think... did you have something else in mind?
![Screenshot 2023-06-07 at 9 47 29 AM](https://github.com/elastic/kibana/assets/17276605/71316eb6-ff07-45b8-9cac-f3388bbde227)

lgestc replied:
this looks okay to me, but maybe @ferenrigue  has some better idea:)

lgestc commented on x-pack/plugins/security_solution/public/flyout/left/components/prevalence_details_prevalence_cell.test.tsx:
```diff
@@ -0,0 +1,171 @@
+
+describe('PrevalenceDetailsAlertCountCell', () => {
+  it('should show loading spinner when useFetchFieldValuePairWithAggregation loading', () => {
+    (useFetchFieldValuePairWithAggregation as jest.Mock).mockReturnValue({
```
jest.mocked thing :) check it out, it's really convenient

lgestc commented on x-pack/plugins/security_solution/public/flyout/left/components/prevalence_details_prevalence_cell.tsx:
```diff
@@ -0,0 +1,84 @@
+
+  const prevalence = aggregationCount / uniqueCount;
+  if (aggregationError || uniqueError || !isFinite(prevalence)) {
+    return null;
```
again, not sure if hiding the error is the way to go

lgestc reviewed (commented):
looking good, one thing is that `undefined` returned inside `LeftFlyoutContext`, see my comment please:)

PhilippeOberti committed:
PR comments:
- remove unnecessary Storybook stories (analyzer graph and session view)
- replace as jest.Mock by jest.mocked where appropriate
- improve useMemo inside PrevalenceDetails
- cell components return error EuiIcon instead of null in case of errors

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/left/context.tsx:
```diff
@@ -73,17 +86,28 @@ export const LeftPanelProvider = ({ id, indexName, children }: LeftPanelProvider
+            searchHit,
+            getFieldsData,
           }
         : undefined,
```
I agree, though unless I'm mistaken this is not introduced in this PR (and is also present in the right context file. Do you mind if I tackle this in a subsequent PR.
I've been wanting to clean this up for a while, as I'm also tired to have to check in all our components if `eventId` or `dataFormattedForFieldBrowser` or `dataAsNestedObject`... are null. There must be a better way to handle this!

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/shared/hooks/use_fetch_field_value_pair_by_event_type.ts:
```diff
@@ -0,0 +1,180 @@
+  /**
+   * Highlighted field value
+   */
+  values: string[];
```
good point on the documentation, I'm now using the same doc as in the cell components.

As for using a single property, I can definitely do that. I wondering though how useful it would be: we're getting the properties from [here](https://github.com/elastic/kibana/pull/158606/files#diff-d5296dcc826179bd0eda3a3c4f2eb94c62d346d2a0b8f5821c37783bdfa871b4R133-R134), so we would be there constructing an object to then deconstruct it here?

lgestc replied:
it would just colocate related things together I guess, if you feel like the updated docs would be enough, just go for it:)

PhilippeOberti replied:
change implemented [here](https://github.com/elastic/kibana/pull/158606/commits/2c8f2af9d3abcef7f75a92b166f5491d7641557f), thanks for the suggestion, I like it!

PhilippeOberti committed:
combine field and values into highlightedField prop

PhilippeOberti committed:
fix after rebase

christineweng commented on x-pack/plugins/security_solution/public/flyout/left/components/translations.ts:
```diff
@@ -85,6 +85,55 @@ export const RELATED_USERS_TOOL_TIP = i18n.translate(
+export const PREVALENCE_TABLE_ALERT_COUNT_COLUMN_TITLE = i18n.translate(
+  'xpack.securitySolution.flyout.prevalenceTableAlertCountColumnTitle',
+  {
+    defaultMessage: 'Alert Count',
```
elastic styles prefers sentence case: i.e. 'Alert count'

PhilippeOberti replied:
yup, good point, fixed!

christineweng reviewed (commented):
Desk tested and LGTM!

PhilippeOberti committed:
sentence case on table headers

PhilippeOberti committed:
fix linting and unit tests
